src/modules/processing/services/nicovideo.js

```diff
@@ -0,0 +1,197 @@
+  }
+}
+
+async function getBasicVideoInformation(id) {
+  const page = await fetch(util.format(NICOVIDEO_EMBED_URL, id), {
+    headers: { "user-agent": genericUserAgent },
+  })
+    .then((response) => response.text())
+    .catch(() => {
+      throw new CobaltError("ErrorCouldntFetch");
+    });
+
+  const data = JSON.parse(
+    page
+      .split('data-props="')
+      .pop()
+      .split('" data-style-map="')
+      .shift()
+      .replaceAll("&quot;", '"')
+  );
+
+  const author = await fetch(
+    util.format(NICOVIDEO_AUTHOR_DATA_URL, data.videoUploaderId),
+    {
+      headers: { "user-agent": genericUserAgent },
+    }
+  )
+    .then((response) => response.json())
+    .catch(() => {
+      throw new CobaltError("ErrorCouldntFetch");
+    });
+
+  return { ...data, author };
+}
+
+async function fetchGuestData(id, actionTrackId) {
+  const data = await fetch(
+    util.format(NICOVIDEO_GUEST_API_URL, id, actionTrackId),
+    {
+      headers: { "user-agent": genericUserAgent },
+    }
+  )
+    .then((response) => response.json())
+    .catch(() => {
+      throw new CobaltError("ErrorCouldntFetch");
+    });
+
+  if (data?.meta?.status !== 200) {
+    throw new CobaltError("ErrorBadFetch");
+  }
+
+  const { videos, audios, accessRightKey } = data.data.media.domand;
+
+  // getting the HQ audio
+  const { id: audioId } = audios
+    .filter((audio) => audio.isAvailable)
+    .sort((firstAudio, secondAudio) => firstAudio.bitrate - secondAudio.bitrate)
+    .shift();
+
+  return {
+    accessRightKey,
+    outputs: videos
+      .filter((video) => video.isAvailable)
+      .map((video) => [video.id, audioId]),
+  };
+}
+
+async function fetchContentURL(id, actionTrackId, accessRightKey, outputs) {
+  const data = await fetch(
+    util.format(NICOVIDEO_HLS_API_URL, id, actionTrackId),
+    {
+      method: "POST",
+      headers: {
+        "user-agent": genericUserAgent,
+        "content-type": "application/json; charset=utf-8",
+        "x-access-right-key": accessRightKey,
+        ...NICOVIDEO_EMBED_FRONTEND_HEADERS,
+      },
+      body: JSON.stringify({ outputs }),
+    }
+  )
+    .then((response) => response.json())
+    .catch(() => {
+      throw new CobaltError("ErrorCouldntFetch");
+    });
+
+  if (data?.meta?.status !== 201) {
+    throw new CobaltError("ErrorBadFetch");
+  }
+
+  return data.data.contentUrl;
+}
+
+async function getHLSContent(contentURL, quality, isAudioOnly, isAudioMuted) {
+  const hls = await fetch(contentURL)
+    .then((response) => response.text())
+    .then((response) => HLS.parse(response));
+
+  const height = quality === "max" ? 9000 : parseInt(quality, 10);
+  let hlsContent = hls.variants.find(
+    (variant) => variant.resolution.height === height
+  );
+
+  if (hlsContent === undefined) {
+    hlsContent = hls.variants
+      .sort(
+        (firstVariant, secondVariant) =>
+          firstVariant.bandwidth - secondVariant.bandwidth
+      )
+      .shift();
```

Comment:
could this be .reduce instead or am i missing something (above too)

Reply:
yep, reduce() fits much better

```javascript
// getting the HQ audio
const { id: audioId } = audios
  .filter((audio) => audio.isAvailable)
  .reduce(
    (firstAudio, secondAudio) =>
      firstAudio.bitRate > secondAudio.bitRate
        ? firstAudio
        : secondAudio
  )
```